src/append_only_data.rs

```diff
@@ -357,14 +370,49 @@ macro_rules! impl_appendable_data {
         #[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash)]
         pub struct $flavour<P>
         where
-            P: Permissions + std::hash::Hash,
+            P: Permissions + std::hash::Hash + Clone,
         {
             inner: AppendOnly<P>,
         }
 
+        impl<P> $flavour<P>
+        where
+            P: Permissions + std::hash::Hash + Clone,
+        {
+            pub fn shell(&self, index: u64) -> Result<Self> {
+                if index > self.entry_index() {
+                    return Err(Error::NoSuchEntry);
+                }
+
+                let mut perm_vec: Vec<P> = Vec::new();
```

Comment:
We could probably just use `filter` here?  Or `take_while` if the indices are always held in increasing order?  E.g.
```rust
let permissions = self
    .inner
    .permissions
    .iter()
    .filter(|perm| perm.data_index() <= index)
    .cloned()
    .collect();
let owners = self
    .inner
    .owners
    .iter()
    .filter(|owner| owner.data_index <= index)
    .cloned()
    .collect();
```